contracts/src/token/erc1155/extensions/supply.rs

```diff
@@ -0,0 +1,204 @@
+//! Extension of ERC-1155 that adds tracking of total supply per id.
+//!
+//! Useful for scenarios where Fungible and Non-fungible tokens have to be
+//! clearly identified. Note: While a totalSupply of 1 might mean the
+//! corresponding is an NFT, there is no guarantees that no other token
+//! with the same id are not going to be minted.
+//!
+//! NOTE: This contract implies a global limit of 2**256 - 1 to the number
+//! of tokens that can be minted.
+//!
+//! CAUTION: This extension should not be added in an upgrade to an already
+//! deployed contract.
+use alloc::vec::Vec;
+
+use alloy_primitives::{uint, Address, U256};
+use stylus_proc::sol_storage;
+use stylus_sdk::prelude::*;
+
+use crate::{
+    token::erc1155::{Erc1155, Error},
+    utils::math::storage::SubAssignUnchecked,
+};
+
+sol_storage! {
+    /// State of [`crate::token::erc1155::Erc1155`] token's supply.
+    pub struct Erc1155Supply {
```

Comment:
I think those extensions `Erc1155Supply` and `Burnable` shouldn't be a part of this pr. Frankly it is already too big. Create a separate branch for each extension. Let's merge pure Erc1155 and it will be soon part of the `0.2.0` release on the next week. We can add extensions later
cc @bidzyyys